scanpy_scripts/lib/_filter.py

```diff
@@ -163,45 +184,48 @@ def _get_filter_conditions(adata, attributes, param, category, subset):
     pct_top = []
     qc_vars_pattern = re.compile(r'^pct_counts_(?P<qc_var>\S+)$')
     qc_vars = []
+
     for name, vmin, vmax in param:
-        if name in ('n_counts', 'index'):
-            logging.error(
-                'Ambiguous parameter name [%s] given, choose from '
-                '"gene:{%s}" or "cell:{%s}"', name, name, name)
-            raise ValueError
-        pt_match = percent_top_pattern.match(name)
-        if pt_match:
-            if name not in adata.obs.columns:
-                pct_top.append(int(pt_match['n']))
-            conditions['cell']['numerical'].append([name, vmin*100, vmax*100])
+        found, cond_cat, cond_name = _attributes_exists(
+            name, attributes, 'numerical')
+        pt_match = percent_top_pattern.match(cond_name)
+        qv_match = qc_vars_pattern.match(cond_name)
+        if found > 1:
+            logging.warning('Parameter "%s" found in both cell and gene '
```

Comment:
Why is this just a warning? Aren't we in danger of silently dropping a filter the user thinks they're applying?

Reply:
If it's specific to this line, then it won't happen in galaxy as all will be prefixed. There are also other warnings for skipping variable or failing to flag MT genes, etc. In general, do you think we should just exit with error messages whenever the run isn't perfect?

Reply:
As a user I would definitely expect that if I specify a filter, then the object I receive as a result has that filter applied (I'm not necessarily monitoring my STDERR). So if something isn't valid for filtering, then I would expect the process to die.

Reply:
Does this apply to all non-perfect situations, e.g. fail to flag MT genes, fail to transfer certain attributes when converting format etc, or do you think it's case by case and you just want it to die in this case? There are situations when users don't necessarily know what features are available in a given object or impractical find out, so opt to just specify some "common" filters to get as close to as possible. Is that a valid reason for keeping them as warnings?

Reply:
From a pipelines point of view I need things to be quite deterministic and predictable. I'd like to know when requested operations are not successful due to errors on my part, so I can fix them. If I specify filtering on an attribute that doesn't exist, in my opinion the process should definitely die, alerting me to my idiocy. If I'm removing mitochondrial genes I need to do a good job of flagging them and the script should fail if I haven't - I'd rather the script didn't make guesses based on 'MT-' for example (I honestly don't thing that's likely to work very reliably).

But that's just our use case of course.

Reply:
Could there be a 'permissive' switch or something to allow operation both ways?

Reply:
Ok, I've changed the behaviour so that it dies when filter variable is unavailable or ambiguous. But lack of MT still only gives a warning. Warning messages are there for a reason, simply ignoring them is not good practice.

Reply:
Thanks @nh3 . Obviously I wouldn't ignore warnings interactively, but running in an automated manner they're hard to catch without some dodgy grepping of STDERR or obsessive checking of logs. Maybe some use of informative exit codes? Anyway, not important right now- thanks for your efforts.